Modules/mathmodule.c

```diff
@@ -3314,6 +3314,34 @@ math_nextafter_impl(PyObject *module, double x, double y)
 }
 
 
+/*[clinic input]
+math.ulp -> double
+
+    x: double
+    /
+
+Return the unit in the last place of x.
+[clinic start generated code]*/
+
+static double
+math_ulp_impl(PyObject *module, double x)
+/*[clinic end generated code: output=f5207867a9384dd4 input=f1004db2d2436ab5]*/
+{
+    if (Py_IS_INFINITY(x) || Py_IS_NAN(x)) {
+        return x;
```

Comment:
It can return negative value.

Reply:
Currently, ulp(-INF) returns -INF. Do you prefer ulp() result to always be positive or NaN? So return +INF for ulp(-INF)?

Reply:
It is the only case when `ulp()` returns a negative value. It looks surprising. And does not confirm the current documentation.

`ulp(nan)` also does not confirm the current documentation (because `nan <= nan` is false). Would not be better to raise an exception?

What different implementations of `ulp()` does? Is there some specification?

Reply:
> It is the only case when ulp() returns a negative value.

Ok, I changed this behavior.

> ulp(nan) also does not confirm the current documentation (because nan <= nan is false). Would not be better to raise an exception?

Hum, should I just document that ulp(x) returns x if x is not a number?

I dislike the exception idea. Other math functions also return NaN if one argument is NaN. For example, nextafter() returns NaN.

Reply:
Java returns +inf for an infinity input, so we'd be consistent with Java. I agree with @serhiy-storchaka that we should return a positive number (or NaN) in all cases.

For NaN, I think it's fine to propagate as usual, but the documentation should specify that that's what happens.

Reply:
done